packages/plugin-response-cache/src/ApolloServerPluginResponseCache.ts

```diff
@@ -0,0 +1,377 @@
+  // requestContext.request.http.headers for a specific authentication header or
+  // cookie.
+  //
+  // This hook may return a promise because, for example, you might need to
+  // validate a cookie against an external service.
+  //
+  // Note: this hook has been updated in Apollo Server v4 to only return a
+  // Promise. This function should always be `await`ed, that way non-TS users
+  // won't experience a breakage (we can await Promises as well as values).
+  sessionId?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): Promise<string | null>;
+
+  // Define this hook if you want the cache key to vary based on some aspect of
+  // the request other than the query document, operation name, variables, and
+  // session ID. For example, responses that include translatable text may want
+  // to return a string derived from
+  // requestContext.request.http.headers.get('Accept-Language'). The data may
+  // be anything that can be JSON-stringified.
+  //
+  // Note: this hook has been updated in Apollo Server v4 to only return a
+  // Promise. This function should always be `await`ed, that way non-TS users
+  // won't experience a breakage (we can await Promises as well as values).
+  extraCacheKeyData?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): Promise<any>;
+
+  // If this hook is defined and returns false, the plugin will not read
+  // responses from the cache.
+  //
+  // Note: this hook has been updated in Apollo Server v4 to only return a
+  // Promise. This function should always be `await`ed, that way non-TS users
+  // won't experience a breakage (we can await Promises as well as values).
+  shouldReadFromCache?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): Promise<boolean>;
+
+  // If this hook is defined and returns false, the plugin will not write the
+  // response to the cache.
+  //
+  // Note: this hook has been updated in Apollo Server v4 to only return a
+  // Promise. This function should always be `await`ed, that way non-TS users
+  // won't experience a breakage (we can await Promises as well as values).
+  shouldWriteToCache?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): Promise<boolean>;
+
+  // This hook allows one to replace the function that is used to create a cache
+  // key. By default, it is the SHA-256 (from the Node `crypto` package) of the result of
+  // calling `JSON.stringify(keyData)`. You can override this to customize the serialization
+  // or the hash, or to make other changes like adding a prefix to keys to allow for
+  // app-specific prefix-based cache invalidation. You may assume that `keyData` is an object
+  // and that all relevant data will be found by the kind of iteration performed by
+  // `JSON.stringify`, but you should not assume anything about the particular fields on
+  // `keyData`.
+  generateCacheKey?(
+    requestContext: GraphQLRequestContext<Record<string, any>>,
+    keyData: unknown,
+  ): string;
```

Comment:
@glasser I know there was some discussion around this possibly being a promise - is that something we _wanted_? This would be the time to make that change if so. I suppose not, I think it was deliberately sync, but wanted to ask to be sure.

Reply:
@kschrade suggested that there wasn't a real use case for this to be async and expressed performance concerns about adding more async hooks (albeit without benchmarks).

My main opinion is I do not like `ValueOrPromise`. I'm fine with Promise or not...